gapis/api/vulkan/framegraph.go

```diff
@@ -166,6 +212,45 @@ type framegraphInfoHelpers struct {
 	rpInfos  []*renderpassInfo
 	rpInfo   *renderpassInfo
 	currRpId uint64
+	// imageLookup is a lookup table to quickly find images that match a memory
+	// access. Scanning the whole state to look for an image matching a memory
+	// access is slow, this lookup table saves seconds of computation. It is
+	// updated when a new renderpass is started under a new top-level parent
+	// command: images can only be created/destroyed between top-level commands,
+	// not during the execution of subcommands.
+	imageLookup  map[memory.PoolID]map[memory.Range][]*ImageObjectʳ
+	parentCmdIdx uint64 // used to know when to update imageLookup
+}
+
+// updateImageLookup updates the lookup table to quickly find an image matching a memory observation.
+func (helpers *framegraphInfoHelpers) updateImageLookup(state *State) {
+	helpers.imageLookup = make(map[memory.PoolID]map[memory.Range][]*ImageObjectʳ)
+	for _, image := range state.Images().All() {
+		for _, aspect := range image.Aspects().All() {
+			for _, layer := range aspect.Layers().All() {
+				for _, level := range layer.Levels().All() {
+					pool := level.Data().Pool()
+					memRange := level.Data().Range()
+					if _, ok := helpers.imageLookup[pool]; !ok {
+						helpers.imageLookup[pool] = make(map[memory.Range][]*ImageObjectʳ)
+					}
+					imgObj := image
+					helpers.imageLookup[pool][memRange] = append(helpers.imageLookup[pool][memRange], &imgObj)
+				}
+			}
+		}
+	}
+}
+
+// lookupImages returns all the images that contain (pool, memRange).
+func (helpers *framegraphInfoHelpers) lookupImages(pool memory.PoolID, memRange memory.Range) []*ImageObjectʳ {
+	images := []*ImageObjectʳ{}
+	for imgRange := range helpers.imageLookup[pool] {
+		if imgRange.Includes(memRange) {
+			images = append(images, helpers.imageLookup[pool][imgRange]...)
```

Comment:
can one memRange included by different memRanges in the map?

Reply:
Here in theory there could be several imgRange that could include memRange, yes. Maybe not in practice, but it doesn't hurt to check them all, since in practice there are almost always only one range associated to a poolID (so the loop on `helpers.imageLookup[pool]` will have only one iteration).